[12.0] [WIP] database_cleanup: migration to 12.0

mart-e opened the pull request:
As part of the [hacktoberfest](http://hacktoberfest.digitalocean.com/), I am starting to migrate the module database_cleanup to 12.0 (hopefully nobody is working on it yet).

- [x] update version number
- [x] apply new README rules
- [x] squash OCA Transbot
- [x] migration scripts
- [x] date(time)
- [x] label/filter
- [x] SASS
- [x] test

Commit:
[ADD] Database cleanup module

Commit:
remove relations when purging models
avoid ''NoneType' object has no attribute 'exists'' error when purging models
fix my change guewen.baconnier@camptocamp.com-20140203103254-v1mzu2uib047xb9h, wrong lines replaced...

Commit:
[FIX] Don't remove uid field from wkf_instance, which is written in
    raw SQL query (but never read afterwards). Workaround for
    lp:1277899

[FIX] Preserve dangling workflow table which is in use

[RFR] Group models per table when detecting columns to purge
      to prevent problems with models sharing the same table

[ADD] Allow purging of dangling data entries

[FIX] Data purging now working

[IMP] Docstrings

[FIX] Label
[FIX] Catch attempt to unlink field from nonexisting model

[RFR] Flake8

anthony-muschang committed:
[CHG] Migration to 8.0

[CHG] database_cleanup: move description to README.rst

hbrunn committed:
[IMP] hide unnecessary buttons in wizard

[IMP] order wizard lines by name

[IMP] deal with modules whose models can't be loaded

[IMP] double quotes for docstring

[FIX] use exists query instead of huge in list

[IMP] hide unnecessary buttons in wizard II

[IMP] readability

[FIX] cope with purging nonexisting models

hbrunn committed:
[IMP] allow to select lines to purge in a tree view

[IMP] adaptto upstream changes

sysadminmatmoz committed:
Missing templates and translations added

hbrunn committed:
[ADD] allow to clean up menus

oca-transbot committed:
OCA Transbot updated translations from Transifex

hbrunn committed:
[ADD] migrate database_cleanup

[ADD] test purging modules

[ADD] test purging tables

hbrunn committed:
[FIX] purge uninstalled uninstallable modules instead of deleting

[FIX] database_cleanup reloads the registry

which has weird side effects during testing. Take care
database_cleanup's tests don't mess up the following tests

StefanRijnhart committed:
[RFR] Explicit access rights so that tests can run
Fixes #505

hbrunn committed:
[IMP] call the button handler to also purge reverse dependencies

hbrunn committed:
[FIX] don't crash if an xmlid refers to a nonexisting field (#559)

JordiBForgeFlow committed:
[FIX] clean database menus was proposing good menus to purge (#562)

florian-dacosta committed:
fixes #587

moylop260 committed:
[FIX] database_cleanup: Fix test (#612)

lasley committed:
[9.0][FIX] database_cleanup: Isolate build (#719)

* [FIX] database_cleanup: Isolate build
* Isolate `database_cleanup` into its own build in Travis file to fix #689

* [FIX] database_cleanup: Remove KeyError assertion
* Remove KeyError assertion in tests due to PR in comment being merged

hbrunn committed:
[ADD] allow creating missing indexes and purging properties (#736)

* [ADD] allow creating missing indexes

* [FIX] tests; installation

* [ADD] allow purging properties

* [ADD] missing file

* [ADD] test purging properties

* [ADD] missing parent_id for menu entry

* [FIX] don't delete too many and wrong properties

eantones committed:
[MIG] database_cleanup: Migration to version 10.0

Using new base model inheritance.

hbrunn committed:
[FIX] really uninstall modules and avoid a crash on cached data

[FIX] don't try to uninstall uninstalled modules

[DEL] weird code

[FIX] actually cleanup where we can

hbrunn committed:
[ADD] [database_cleanup] migrate to 11.0

hbrunn committed:
[IMP] travis: isolate database_cleanup again

Cherry-pick of d464fb2c596225 to 12.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

fanha99 committed:
database_cleanup_fix_runbot_links

simahawk committed:
Global pylint cleanup

mart-e committed:
[MIG] database_cleanup: migration to 12.0

Update version number
Use correct import
Update .pot to match new syntax
Adapt icon to fontawesome
Add models description
Rename field external id to new syntax <model>__<field>

pedrobaeza commented:
Great! No, nobody is working on this yet, so I assign you in #1368

mart-e commented:
@pedrobaeza should be good now !

gustavovalverde commented:
Teste with a module migration use case from V11, which was also done in v12. Almost all seems like working OK, except for a weird behavior which I couldn't reproduce again when Purging all columns; after doing that it started complaining that my user (admin 1) didn't have enough priviliges when clicking on all menus. After a refresh a *403 Forbidden* was raised.

After restoring a snapshot and doing the process again, this did not happened; and everything seems working as expected.

gustavovalverde commented:
Found the culprit; it's the `password` column from `res.users`. If purged it locks you out of the system. cc: @mart-e @pedrobaeza

hbrunn committed:
[FIX] don't destroy values when there are empty default properties

hbrunn committed:
[FIX] show the redundant property's res_id, not the default one's

hbrunn committed:
[IMP] don't break on properties pointing to unknown models

paulius-sladkevicius commented:
I've tested it out on the v12, works well. However I found `password` and `password_crypt` in the list of columns that wants to purge, so @gustavovalverde is right.

pedrobaeza commented:
#1507 should be included here.

bealdav commented:
@pedrobaeza  Don't you think we can merge and add a comment in roadmap to cherry pick v11 changes .

gustavovalverde commented:
@bealdav this shouldn't be merged until the passwords issue is solved (at least): https://github.com/OCA/server-tools/pull/1408#issuecomment-443593754 

P.D.: I'm using it already, with this PR, knowing that I have to remove the password table from being purged. But someone can have a really bad moment not knowing this.

pedrobaeza commented:
Well, that comment means to not do it ever, and that changes are very important, even possibly solving @gustavovalverde's issue. If you want to retake the migration and add that PR, it's very straight-forward.

pedrobaeza commented:
@mart-e has just pushed an updated branch. Please check your issue, @gustavovalverde

mart-e commented:
Hi,

Sorry for the lack of replies. I will try to fix the remaining issue.

pedrobaeza commented:
Thanks!

mart-e committed:
[FIX] database_cleanup: exclude password fields

cf discussion on OCA/server-tools#1408

mart-e commented:
I have also picked the commits from #1507, I hope it was the only ones.

gustavovalverde commented:
@bealdav and @paulius-sladkevicius could you guys check and approve if everything is working as expected?

mart-e committed:
[I18N] database_cleanup: reflect changes to en.po

To match the new database_cleanup.pot

mart-e commented:
By the way, the translation files from the previous version are outdated. I have updated en.po at mart-e/server-tools@465c6e4561b1e136842e7 for the example but should I apply it everywhere or let the translation memory do its magic?

pedrobaeza commented:
@mart-e it seems latest changes have broken Travis

mart-e committed:
[FIX] database_cleanup: filter on many2many fields

Following odoo/odoo@54238c88e85d6e a Many2many may have no relation
This avoids having a None record in the list (which is problematic with the
NOT IN clause)
https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_NOT_IN

mart-e commented:
@pedrobaeza got it, the culprit was odoo/odoo@54238c8 which is problematic when combined with a `NOT IN` in SQL
https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_NOT_IN

pedrobaeza commented:
@mart-e still no luck I'm afraid, but seeing logs I'm not sure why it's failing.

mart-e committed:
[FIX] database_cleanup: ignore to_buy modules

Since 12.0, some virtual modules are created as an incentive to buy enterprise
Hide installation warning

mart-e commented:
> but seeing logs I'm not sure why it's failing.

Because we had the good idea to introduce fake modules with a `to_buy` flag in 12.0 :roll_eyes: 
However, what I am still not sure is about this one:

https://github.com/mart-e/server-tools/blob/8fb4995ce35ff8080c899c7af14fbd5a2da56ccc/database_cleanup/models/purge_modules.py#L50-L52

when purging modules, you try to uninstall it first but this will produce a warning as the created module is not a real one.
I am not really sure what should be the right behaviour when purging modules.

pedrobaeza commented:
> However, what I am still not sure is about this one:
> https://github.com/mart-e/server-tools/blob/8fb4995ce35ff8080c899c7af14fbd5a2da56ccc/database_cleanup/models/purge_modules.py#L50-L52
> when purging modules, you try to uninstall it first but this will produce a warning as the created module is not a real one.
> I am not really sure what should be the right behaviour when purging modules.

This was introduced in 33727c5ae709cb3e9d68ec9ba1168a0c83978126 by @hbrunn.

I think the warning is not a problem, but these 2 errors:
```
2019-05-09 15:33:41,645 6556 ERROR openerp_test odoo.modules.registry: Model base.exception.test.purchase has no table.
2019-05-09 15:33:41,645 6556 ERROR openerp_test odoo.modules.registry: Model base.exception.test.purchase.line has no table.
```
They are not in the main branch, so it should be somehow provoked due to this module. However, I have restarted build on main branch to see if something new since last build due to any change upstream:

https://travis-ci.org/OCA/server-tools/jobs/529353188

pedrobaeza commented:
Main server-tools branch is green after testing again, so definitively something here is provoking that errors.

mart-e commented:
@hbrunn do you recall the context of d464fb2c596225f91b3968e4643138a449cf92db ? I think I got the same issue with 12.0 migration, incompatibility between tests of base_exception and database_cleanup

hbrunn commented:
the thing is that some of the actions this module calls involve opening a new cursor and committing transactions, so the database will contain garbage after the tests

pedrobaeza commented on .travis.yml:
```diff
@@ -17,9 +17,11 @@ env:
-  - TESTS="1" ODOO_REPO="OCA/OCB"
+  - TESTS="1" ODOO_REPO="OCA/OCB" EXCLUDE="database_cleanup"
+  - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="database_cleanup"
   - TESTS="1" ODOO_REPO="OCA/OCB" MAKEPOT="1"
```
I see that we are not testing against "odoo/odoo" and this line is also including both together, which makes this to fail. Please put:
```
- TESTS="1" ODOO_REPO="odoo/odoo" EXCLUDE="database_cleanup" MAKEPOT="1"
- TESTS="1" ODOO_REPO="odoo/odoo" INCLUDE="database_cleanup" MAKEPOT="1"
```

The MAKEPOT part is for synchronizing translations with weblate.

pedrobaeza commented:
Only one flake8 error to go:
```
database_cleanup/models/purge_modules.py:66:80: E501 line too long (84 > 79 characters)
```
We use strict PEP8 80 cols limit.

And with this and the other change, everything will be finally ready. Thanks for the efforts.

As you are in holidays this week, I'll do it for you for finishing the PR.

pedrobaeza commented:
Merging after having green. Thanks for the efforts done!

mart-e commented:
Thanks for the final cleanup, glad this was finally merged !
